### PR TITLE
Added option to only run the compressor on a range of state groups in a room

### PR DESCRIPTION
**Pull request by Azrenbeth**
BUILDS ON azren/added_graph_option (PR #43)

This allows the compressor to be called on only a selection of state groups in a room.

-b can be used to select the state group number to start compressing from 
(entering 0 or omitting this will start at the first state_group in the room)

-n can be used to enter (roughly) how many state groups you wish to compress. 

The compressor will print to stdout where it got up to. Entering this state group number
on the next run of the compressor will start from where the previous run finished

(NOTE: chunking produces worse compression than running all in one go, but will save memory)

### CODE REVIEW

**Comment by erikjohnston** [comment]
> NOTE - changes meaning of -s option from last state group to compress to first state group

Why are we removing the option? I'd much rather we kept it and used a different name for the new option?

File: src/database.rs [hunk 1]
```diff
@@ -76,41 +89,85 @@ pub fn get_data_from_db(
...
+    let rows = if let (Some(min), Some(count)) = (min_state_group, groups_to_compress) {
+        let params: Vec<&dyn ToSql> = vec![&room_id, &min, &count];
+        client.query_raw(format!(r"{} AND m.id > $2 LIMIT $3", sql).as_str(), params)
+    } else {
+        client.query_raw(sql, &[room_id])
+    }
+    .unwrap();
+
+    let final_row = rows.last().unwrap().unwrap();
+    final_row.get(0)
```
Comment (erikjohnston) [inline src/database.rs:146]: If we're only interested in the last row then we should avoid pulling all the rows out of the DB, as that might be many many rows and a lot of wasted effort. When `min_state_group` is None then we can do that via `ORDER BY id DESC LIMIT 1`

File: src/lib.rs [hunk 2]
```diff
@@ -70,13 +71,39 @@ impl FromStr for LevelSizes {
...
+    // If the compressor results in less than this many rows being saved then
+    // it will abort
     min_saved_rows: Option<i32>,
-    transactions: bool,
+    // The sizes of the different levels in the new state_group tree being built
     level_sizes: LevelSizes,
+    // Whether or not to wrap each change to an individual state_group in a transaction
+    // This is very much reccomended when running the compression when synapse is live
+    // TODO: should this actually be an opt-out flag? it's much worse to need it and
+    //       forget to add it than to not need it and forget to remove it....?
```
Comment (erikjohnston) [inline src/lib.rs:103]: It's not needed if the script itself is run in one big tranasction?

File: src/compressor.rs [hunk 3]
```diff
@@ -141,6 +141,23 @@ impl<'a> Compressor<'a> {
         pb.enable_steady_tick(100);
 
         for (&state_group, entry) in self.original_state_map {
+            // Check whether this entry is in_range or is just present in the map due to being
+            // a predecessor of a group that IS in_range for compression
+            if !entry.in_range {
```
Comment (erikjohnston) [inline src/compressor.rs:146]: We don't ever seem to set this to false anywhere?

**Review by erikjohnston — CHANGES_REQUESTED** [review]

File: src/compressor.rs (hunk 3 above)
Comment (Azrenbeth) [reply src/compressor.rs:146]: we create StateGroupEntries with default() which sets it to false, and then it gets explicity sets to true for some state groups

Comment (Azrenbeth) [reply src/compressor.rs:146]: It's only the state groups loaded with get_missing_from_db that can not be in range (i.e. groups just loaded because they're the predecessors of groups that are in range)

File: src/database.rs [hunk 4]
```diff
@@ -76,41 +89,85 @@ pub fn get_data_from_db(
...
+    room_id: &str,
+    min_state_group: Option<i64>,
+    groups_to_compress: Option<i64>,
+) -> i64 {
+    // Get list of state_id's in a certain room
+    let sql = r#"
+        SELECT m.id
+        FROM state_groups AS m
+        WHERE m.room_id = $1
+    "#;
```
[the remark this answers is not in the document]
Comment (Azrenbeth) [reply src/database.rs]: done

File: src/database.rs (hunk 1 above)
Comment (Azrenbeth) [reply src/database.rs:145]: done

File: src/database.rs [hunk 5]
```diff
@@ -76,41 +89,85 @@ pub fn get_data_from_db(
...
+
+    // Adds additional constraint if a groups_to_compress has been specified
+    // Then sends query to the datatbase
+    let rows = if let (Some(min), Some(count)) = (min_state_group, groups_to_compress) {
+        let params: Vec<&dyn ToSql> = vec![&room_id, &min, &count];
+        client.query_raw(format!(r"{} AND m.id > $2 LIMIT $3", sql).as_str(), params)
```
[the remark this answers is not in the document]
Comment (Azrenbeth) [reply src/database.rs]: done

**Comment by Azrenbeth** [comment]
> > NOTE - changes meaning of -s option from last state group to compress to first state group
> 
> Why are we removing the option? I'd much rather we kept it and used a different name for the new option? That way we retain backwards compat.

I removed it because I thought "max_state_group" was there just to limit the number of groups you work on at once, which the chunking options replace with something better, and removing it simplified the code a bit

(But mainly I just forgot to think about backwards compatability...)

File: src/compressor.rs (hunk 3 above)
Comment (erikjohnston) [reply src/compressor.rs:146]: Aha! Makes sense, thanks

File: src/database.rs [hunk 6]
```diff
@@ -76,41 +98,90 @@ pub fn get_data_from_db(
...
-        println!("Missing {} state groups", missing_sgs.len());
+        // println!("Missing {} state groups", missing_sgs.len());
 
-        let map = get_missing_from_db(&mut client, &missing_sgs);
-        state_group_map.extend(map.into_iter());
+        // find state groups not picked up already and add them to the map
+        let map = get_missing_from_db(&mut client, &missing_sgs, min_state_group, max_group_found);
+        for (k, v) in map.into_iter() {
+            state_group_map.entry(k).or_insert(v);
+        }
```
Comment (erikjohnston) [inline src/database.rs:114]: Why this change? Also `.into_iter()` is implicit in for loops, so you can do that as `for (k, v) in map { .. }`

File: src/database.rs [hunk 7]
```diff
@@ -76,41 +98,90 @@ pub fn get_data_from_db(
...
+
+    if let Some(max) = max_state_group {
+        sql = format!("{} AND id <= {}", sql, max)
+    }
+
+    // Adds additional constraint if a groups_to_compress has been specified
+    // Then sends query to the datatbase
+    let rows = if let (Some(min), Some(count)) = (min_state_group, groups_to_compress) {
+        let params: Vec<&dyn ToSql> = vec![&room_id, &min, &count];
+        client.query_raw(format!(r"{} AND id > $2 LIMIT $3", sql).as_str(), params)
```
Comment (erikjohnston) [inline src/database.rs:150]:
```suggestion
        client.query_raw(format!(r"{} AND id > $2 OFFSET $3 LIMIT 1", sql).as_str(), params)
```

I think works?

File: src/database.rs [hunk 8]
```diff
@@ -76,41 +98,90 @@ pub fn get_data_from_db(
...
+    } else {
+        client.query_raw(
+            format!(r"{} ORDER BY id DESC LIMIT 1", sql).as_str(),
```
Comment (erikjohnston) [inline src/database.rs:153]:
```suggestion
            &format!(r"{} ORDER BY id DESC LIMIT 1", sql),
```

Does that not work?

File: src/database.rs [hunk 9]
```diff
@@ -76,41 +98,90 @@ pub fn get_data_from_db(
...
+        sql = format!("{} AND id <= {}", sql, max)
+    }
+
+    // Adds additional constraint if a groups_to_compress has been specified
+    // Then sends query to the datatbase
+    let rows = if let (Some(min), Some(count)) = (min_state_group, groups_to_compress) {
+        let params: Vec<&dyn ToSql> = vec![&room_id, &min, &count];
+        client.query_raw(format!(r"{} AND id > $2 LIMIT $3", sql).as_str(), params)
+    } else {
+        client.query_raw(
```
Comment (erikjohnston) [inline src/database.rs:152]: It's probably easier to use `client.query(`?

File: src/database.rs (hunk 6 above)
Comment (Azrenbeth) [reply src/database.rs:114]: I did this so that state groups you get from get_missing_from_db cannot overwrite the entries in state_group_map, which they could before. 

Unless something goes wrong, that should never happen atm. so it was just for peace of mind when I was editing get_missing_from_db (I guess it could be a panic if it was trying to replace anything? As would mean compressor potentially unsafe)

File: src/database.rs (hunk 7 above)
Comment (Azrenbeth) [reply src/database.rs:150]: I think it should be "OFFSET [count-1] LIMIT 1" but yes that works

File: src/database.rs (hunk 8 above)
Comment (Azrenbeth) [reply src/database.rs:153]: For some reason - no... 

One & gives: `ToStatement` is not implemented for `String`
Two & gives: `ToStatement` is not implemented for `&String`

File: src/database.rs (hunk 6 above)
Comment (erikjohnston) [reply src/database.rs:114]: Ah, fair.

File: src/database.rs (hunk 8 above)
Comment (erikjohnston) [reply src/database.rs:153]: Bleurgh, tedious

File: src/database.rs (hunk 9 above)
Comment (Azrenbeth) [reply src/database.rs:152]: well probably... but one of the quirks of the postgres crate is its mysterious type restrictions, so to use OFFSET [count], the variable groups_to_compress will have to be of type Option<i32> not Option<i64>, or you get the errror: 

`Error { kind: ToSql(2), cause: Some(WrongType { postgres: Int4, rust: "i64" }) }'`

There's no particular reason not to just use an i32, but since it avoids hitting this confusing bug, I feel like we should just stick with query_raw

File: src/database.rs (hunk 7 above)
Comment (Azrenbeth) [reply src/database.rs:150]: Hmmm I'm now getting an annoying error from postgres: `WrongType { postgres: Int4, rust: "i64" }`  which seems to be solved by changing groups_to_compress to an Option\<i32\>

Comment (Azrenbeth) [reply src/database.rs:150]: I see no particular reason not to just make it i32 though (because if you want to compress more than the 32 bit limit in one go then that's quite a special case! - and would be better done with the saving and restoring stuff I add in later)

Comment (Azrenbeth) [reply src/database.rs:150]: Hmm There's an issue with OFFSET preventing you from compressing the last chunk of a room

Previously, if there were 400 groups between min_state_group and the highest in the table and you had groups_to_compress as 2000, then it would return the 400th, but with OFFSET, the query would return nothing

File: src/database.rs (hunk 9 above)
Comment (Azrenbeth) [reply src/database.rs:152]: yes, but only if we do use offset

**Review by erikjohnston — APPROVED** [review]
woot

File: src/database.rs (hunk 7 above)
Comment (Azrenbeth) [reply src/database.rs:150]: This works: `SELECT id from (SELECT id FROM state_groups WHERE $1 AND id > $2 LIMIT $3) AS ids ORDER BY ids.id desc LIMIT 1;` 

Is this any worse than doing offset?

Comment (erikjohnston) [reply src/database.rs:150]:
```sql
synapse=# EXPLAIN SELECT id from (SELECT id FROM state_groups WHERE id > 100 LIMIT 1000) AS ids ORDER BY id desc LIMIT 1;
                                                      QUERY PLAN                                                       
-----------------------------------------------------------------------------------------------------------------------
 Limit  (cost=44.78..44.78 rows=1 width=8)
   ->  Sort  (cost=44.78..47.28 rows=1000 width=8)
         Sort Key: state_groups.id DESC
         ->  Limit  (cost=0.43..29.78 rows=1000 width=8)
               ->  Index Only Scan using state_groups_pkey on state_groups  (cost=0.43..66734.45 rows=2273722 width=8)
                     Index Cond: (id > 100)
(6 rows)

```

Looks fine, basically it scans the index starting from the given ID, sorts it, and then takes the last one. It's not quite as efficient as offset in that it accumulates all the ids into a list, but this is a lot better than returning the list to the app :slightly_smiling_face:

Comment (Azrenbeth) [reply src/database.rs]: done

File: src/database.rs (hunk 9 above)
Comment (Azrenbeth) [reply src/database.rs]: done

**Review by erikjohnston — APPROVED** [review]
